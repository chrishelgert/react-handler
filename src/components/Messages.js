import React, { PropTypes } from 'react';

export const LoadingMessage = () => (
  <div className="handler--loading">loading...</div>
);

export const ErrorMessage = ({ children }) => (
  <div className="handler--error">{children}</div>
);

ErrorMessage.propTypes = {
  children: PropTypes.string.isRequired,
};

export const EmptyMessage = ({ children }) => (
  <div className="handler--empty">{children}</div>
);

EmptyMessage.propTypes = {
  children: PropTypes.string.isRequired,
};
