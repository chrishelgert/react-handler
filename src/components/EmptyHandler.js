import React, { PropTypes } from 'react';
import { EmptyMessage } from './Messages';
import isEmpty from '../utils/isEmpty';

/**
 * This Mini-Handler handles only the empty-state
 * or display the given Component.
 */
const EmptyHandler = ({
  emptyMsg,
  emptyPropertyKey,
  EmptyComponent,
  Component,
  componentProps = {},
}) => {
  if (isEmpty(emptyPropertyKey, componentProps)) {
    if (EmptyComponent) {
      return <EmptyComponent>{emptyMsg}</EmptyComponent>;
    }

    return <EmptyMessage>{emptyMsg}</EmptyMessage>;
  }

  return <Component {...componentProps} />;
};

EmptyHandler.propTypes = {
  emptyMsg: PropTypes.string.isRequired,
  emptyPropertyKey: PropTypes.string,
  EmptyComponent: PropTypes.func,
  Component: PropTypes.func.isRequired,
  componentProps: PropTypes.object,
};

export default EmptyHandler;
