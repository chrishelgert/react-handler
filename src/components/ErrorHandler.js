import React, { PropTypes } from 'react';
import { ErrorMessage } from './Messages';

/**
 * This Mini-Handler handles only the error-state
 * or display the given Component.
 */
const ErrorHandler = ({ errorMsg, ErrorComponent, Component, componentProps = {} }) => {
  if (errorMsg) {
    if (ErrorComponent) {
      return <ErrorComponent>{errorMsg}</ErrorComponent>;
    }

    return <ErrorMessage>{errorMsg}</ErrorMessage>;
  }

  return <Component {...componentProps} />;
};

ErrorHandler.propTypes = {
  errorMsg: PropTypes.string,
  ErrorComponent: PropTypes.func,
  Component: PropTypes.func.isRequired,
  componentProps: PropTypes.object,
};

export default ErrorHandler;
