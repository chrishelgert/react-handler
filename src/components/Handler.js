import React, { PropTypes } from 'react';
import { LoadingMessage, ErrorMessage, EmptyMessage } from './Messages';
import isEmpty from '../utils/isEmpty';

/**
 * The Handler handles loading, error and empty state
 * or display the given Component.
 */
const Handler = ({
  LoadingComponent,
  ErrorComponent,
  EmptyComponent,
  Component,
  errorMsg,
  emptyMsg,
  emptyPropertyKey,
  componentProps = {},
  loading = true,
}) => {
  if (loading) {
    return LoadingComponent ? <LoadingComponent /> : <LoadingMessage />;
  }

  if (errorMsg) {
    if (ErrorComponent) {
      return <ErrorComponent>{errorMsg}</ErrorComponent>;
    }

    return <ErrorMessage>{errorMsg}</ErrorMessage>;
  }

  if (isEmpty(emptyPropertyKey, componentProps)) {
    if (EmptyComponent) {
      return <EmptyComponent>{emptyMsg}</EmptyComponent>;
    }

    return <EmptyMessage>{emptyMsg}</EmptyMessage>;
  }

  return <Component {...componentProps} />;
};

Handler.propTypes = {
  LoadingComponent: PropTypes.func,
  ErrorComponent: PropTypes.func,
  EmptyComponent: PropTypes.func,
  Component: PropTypes.func.isRequired,
  errorMsg: PropTypes.string,
  emptyMsg: PropTypes.string,
  emptyPropertyKey: PropTypes.string,
  componentProps: PropTypes.object,
  loading: PropTypes.bool,
};

export default Handler;
