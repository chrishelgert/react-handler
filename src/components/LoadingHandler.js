import React, { PropTypes } from 'react';
import { LoadingMessage } from './Messages';

/**
 * This Mini-Handler handles only the loading-state
 * or display the given Component.
 */
const LoadingHandler = ({
  LoadingComponent,
  Component,
  componentProps = {},
  loading = true,
}) => {
  if (loading) {
    return LoadingComponent ? <LoadingComponent /> : <LoadingMessage />;
  }

  return <Component {...componentProps} />;
};

LoadingHandler.propTypes = {
  LoadingComponent: PropTypes.func,
  Component: PropTypes.func.isRequired,
  componentProps: PropTypes.object,
  loading: PropTypes.bool,
};

export default LoadingHandler;
