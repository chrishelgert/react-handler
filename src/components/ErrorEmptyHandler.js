import React, { PropTypes } from 'react'
import EmptyHandler from './EmptyHandler'
import ErrorHandler from './ErrorHandler'

/**
 * This Sub-Handler combines the Error and Empty-Handler
 */
const ErrorEmptyHandler = ({
  errorMessage,
  ErrorComponent,
  emptyMessage,
  checkedProperty,
  EmptyComponent,
  children
}) => (
  <ErrorHandler message={errorMessage} ErrorComponent={ErrorComponent}>
    <EmptyHandler
      message={emptyMessage}
      checkedProperty={checkedProperty}
      EmptyComponent={EmptyComponent}
    >
      {children}
    </EmptyHandler>
  </ErrorHandler>
)

ErrorEmptyHandler.propTypes = {
  ErrorComponent: PropTypes.func,
  EmptyComponent: PropTypes.func,
  errorMessage: PropTypes.string,
  emptyMessage: PropTypes.string.isRequired,
  checkedProperty: PropTypes.any,
  children: PropTypes.node.isRequired
}

ErrorEmptyHandler.defaultProps = {
  ErrorComponent: null,
  EmptyComponent: null,
  errorMessage: null,
  checkedProperty: null
}

export default ErrorEmptyHandler
