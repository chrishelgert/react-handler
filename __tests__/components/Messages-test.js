import React from 'react';
import renderer from 'react-test-renderer';
import { LoadingMessage, EmptyMessage, ErrorMessage } from '../../src/components/Messages';

describe('Messages', () => {
  describe('LoadingMessage', () => {
    it('should match the snapshot', () => {
      const tree = renderer.create(<LoadingMessage />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('EmptyMessage', () => {
    it('should match the snapshot', () => {
      const tree = renderer.create(<EmptyMessage>empty</EmptyMessage>).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('ErrorMessage', () => {
    it('should match the snapshot', () => {
      const tree = renderer.create(<ErrorMessage>error</ErrorMessage>).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
