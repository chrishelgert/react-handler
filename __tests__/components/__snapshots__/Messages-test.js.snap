exports[`Messages EmptyMessage should match the snapshot 1`] = `
<div
  className="handler--empty">
  empty
</div>
`;

exports[`Messages ErrorMessage should match the snapshot 1`] = `
<div
  className="handler--error">
  error
</div>
`;

exports[`Messages LoadingMessage should match the snapshot 1`] = `
<div
  className="handler--loading">
  loading...
</div>
`;
